### PR TITLE
De-circularize Colorbar and Colorscale modules

### PR DESCRIPTION
**Pull request by etpinard**
@cldougl @mdtusz @alexcjohnson 

Before this PR `Colorbar` and `Colorscale` had a circular dependencies.

By breaking both modules into smaller 1-file-per-method pieces, both modules can be required without any issues.

Most of this PR only moved content around. *Real* changes are highlighted below.

### CODE REVIEW

File: src/components/colorbar/draw.js [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+'use strict';
+
+var d3 = require('d3');
+
+var Plotly = require('../../plotly');
+
+var attributes = require('./attributes');
+
+
+module.exports = function draw(gd, id) {
```
Comment (etpinard) [inline src/components/colorbar/draw.js]: `Plotly.Colorbar` used to be a function with `.attributes` and `.supplyDefaults` attached to it which was rather confusing.

Now, the main `Colorbar` method is call `draw`.

File: src/components/colorbar/has_colorbar.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+
+module.exports = function hasColorbar(container) {
```
Comment (etpinard) [inline src/components/colorbar/has_colorbar.js]: this used to be part of `Plotly.Colorscale`. Makes more sense to put it in `Colorbar`.

File: src/components/colorscale/default_scale.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+var scales = require('./scales');
+
+
+module.exports = scales.RdBu;
```
Comment (etpinard) [inline src/components/colorscale/default_scale.js]: dumb but :palm_tree:

Comment (etpinard) [inline src/components/colorscale/default_scale.js]: .. as the default scale is used in two methods.

File: src/lib/coerce.js [hunk 4]
```diff
@@ -140,7 +140,7 @@ exports.valObjects = {
         requiredOpts: [],
         otherOpts: ['dflt'],
         coerceFunction: function(v, propOut, dflt) {
-            propOut.set(Plotly.Colorscale.getScale(v, dflt));
+            propOut.set(getColorscale(v, dflt));
```
Comment (etpinard) [inline src/lib/coerce.js]: require the method directly instead of relying on `Plotly`

File: src/components/colorbar/has_colorbar.js (hunk 2 above)
Comment (cldougl) [inline src/components/colorbar/has_colorbar.js]: :+1:

**Comment by cldougl** [comment]
Looks good to me. Thanks for highlighting & commenting on *real* changes @etpinard ! :dancer:

**Comment by mdtusz** [comment]
Oh man, these will make huge improvements for read/maintainability. :+1:

**Pull request merged**